crates/bevy_render/macros/src/as_bind_group.rs

```diff
@@ -0,0 +1,389 @@
+        uniform_fields: Vec<&'a Field>,
+    },
+}
+
+pub fn derive_as_bind_group(input: TokenStream) -> TokenStream {
+    let ast = parse_macro_input!(input as DeriveInput);
+
+    let manifest = BevyManifest::default();
+    let render_path = manifest.get_path("bevy_render");
+    let asset_path = manifest.get_path("bevy_asset");
+
+    let mut binding_states: Vec<BindingState> = Vec::new();
+    let mut binding_impls = Vec::new();
+    let mut bind_group_entries = Vec::new();
+    let mut binding_layouts = Vec::new();
+    let mut attr_prepared_data_ident = None;
+
+    // Read struct-level attributes
+    for attr in &ast.attrs {
+        if let Some(attr_ident) = attr.path.get_ident() {
+            if attr_ident == BIND_GROUP_DATA_ATTRIBUTE_NAME {
+                if let Ok(prepared_data_ident) =
+                    attr.parse_args_with(|input: ParseStream| input.parse::<Ident>())
+                {
+                    attr_prepared_data_ident = Some(prepared_data_ident);
+                }
+            } else if attr_ident == UNIFORM_ATTRIBUTE_NAME {
+                let (binding_index, converted_shader_type) = attr
+                    .parse_args_with(|input: ParseStream| {
+                        let binding_index = input
+                            .parse::<LitInt>()
+                            .and_then(|i| i.base10_parse::<u32>())?;
+                        input.parse::<Comma>()?;
+                        let converted_shader_type = input.parse::<Ident>()?;
+                        Ok((binding_index, converted_shader_type))
+                    })
+                    .unwrap_or_else(|_| {
+                        panic!("struct-level uniform bindings must be in the format: uniform(BINDING_INDEX, ConvertedShaderType)");
+                    });
+
+                binding_impls.push(quote! {{
+                    use #render_path::render_resource::AsBindGroupShaderType;
+                    let mut buffer = #render_path::render_resource::encase::UniformBuffer::new(Vec::new());
+                    let converted: #converted_shader_type = self.as_bind_group_shader_type(images);
+                    buffer.write(&converted).unwrap();
+                    #render_path::render_resource::OwnedBindingResource::Buffer(render_device.create_buffer_with_data(
+                        &#render_path::render_resource::BufferInitDescriptor {
+                            label: None,
+                            usage: #render_path::render_resource::BufferUsages::COPY_DST | #render_path::render_resource::BufferUsages::UNIFORM,
+                            contents: buffer.as_ref(),
+                        },
+                    ))
+                }});
+
+                binding_layouts.push(quote!{
+                    #render_path::render_resource::BindGroupLayoutEntry {
+                        binding: #binding_index,
+                        visibility: #render_path::render_resource::ShaderStages::FRAGMENT,
```

Comment:
So, if I want to have a binding visible in the Vertex stage I need to go back to implementing all of this without the macro? I guess this isn't needed nearly as often so it's not a big issue, but I mostly want to know that I am understanding this correctly.

Reply:
I'm asking because it isn't mentioned in the Future Works section

Reply:
Ooops! This was a mistake. I meant to go through and make these all visible in both contexts by default (with a followup item to allow binding derive attributes to define more granular visibility when needed). Forgot to add it to my "pre-PR todo list" and promptly forgot to do it.

Thanks for catching this ... I probably wouldn't have remembered to get back to it prior to merging.

Reply:
I think some bindings are not allowed in the vertex stage. I want to say some texture bindings but I don't remember. It's fuzzy, but something is poking at me about it. :)